File: core/blas/cgerc.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */
#include "cgerc.h"
#include <complex>
#include "blas_common.h"
#include "blasplan/include/blasplan/BlasCgercPlan.h"

using namespace AsdSip;
using namespace Mki;

namespace AsdSip {
struct CgercTensorParam {
    const aclTensor *x;
    const aclTensor *y;
    const aclTensor *A;
};

AspbStatus CgercDtypeCheck(struct CgercTensorParam parm)
{
    auto ret = ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE;
    SIP_OP_CHECK_DTYPE_NOT_MATCH(parm.A, aclDataType::ACL_COMPLEX64, ret);
    SIP_OP_CHECK_DTYPE_NOT_MATCH(parm.x, aclDataType::ACL_COMPLEX64, ret);
    SIP_OP_CHECK_DTYPE_NOT_MATCH(parm.y, aclDataType::ACL_COMPLEX64, ret);
    return ErrorType::ACL_SUCCESS;
}

AspbStatus CgercShapeCheck(struct CgercTensorParam parm, const int64_t m, const int64_t n)
{
    ASDSIP_ECHECK(m > 0 && n > 0 && m <= UINT32_MAX && n <= UINT32_MAX,
        "blas Cgerc get m <= 0 || n <= 0 or m or n > 2^32.",
        ErrorType::ACL_ERROR_INVALID_PARAM);
    SIP_OP_CHECK_NUM_NOT_MATCH(parm.A, m * n, ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH);
    SIP_OP_CHECK_NUM_NOT_MATCH(parm.x, m, ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH);
    SIP_OP_CHECK_NUM_NOT_MATCH(parm.y, n, ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH);
    return ErrorType::ACL_SUCCESS;
}

AspbStatus asdBlasCgerc(asdBlasHandle handle, const int64_t m, const int64_t n, const std::complex<float> &alpha,
    aclTensor *x, const int64_t incx, aclTensor *y, const int64_t incy, aclTensor *A, const int64_t lda)
{
    std::lock_guard<std::mutex> lock(blas_mtx);
    ASDSIP_ECHECK(BlasPlanCache::doesPlanExist(handle),
        "blas Cgerc get cached plan failed.",
        ErrorType::ACL_ERROR_INTERNAL_ERROR);

    struct CgercTensorParam parm {
        x, y, A
    };
    ASDSIP_CHECK(CgercDtypeCheck(parm) == ErrorType::ACL_SUCCESS,
        "blas asdBlasCgerc dtype check failed.",
        return ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE);
    ASDSIP_CHECK(CgercShapeCheck(parm, m, n) == ErrorType::ACL_SUCCESS,
        "blas asdBlasCgerc shape check failed.",
        return ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH);

    if (incx != 1) {
        ASDSIP_LOG(INFO) << "blas asdBlasCgerc get incx != 1.";
    }
    if (incy != 1) {
        ASDSIP_LOG(INFO) << "blas asdBlasCgerc get incy != 1.";
    }
    if (lda <= 0) {
        ASDSIP_LOG(ERROR) << "blas asdBlasCgerc get lda <= 0.";
    }

    try {
        AsdSip::BlasCgercPlan &plan = dynamic_cast<AsdSip::BlasCgercPlan &>(BlasPlanCache::getPlan(handle));
        ASDSIP_ECHECK(plan.IsInitialized(), "Cgerc plan init Error!.", ErrorType::ACL_ERROR_INTERNAL_ERROR);

        OpDesc opDesc;
        opDesc.opName = "CgercOperation";
        AsdSip::OpParam::Cgerc param;
        param.m = m;
        param.n = n;
        param.incx = static_cast<uint32_t>(incx);
        param.incy = static_cast<uint32_t>(incy);
        param.alphaReal = alpha.real();
        param.alphaImag = alpha.imag();
        opDesc.specificParam = param;
        ASDSIP_LOG(DEBUG) << "OpDesc: " << opDesc.opName << "; OpDesc info: " << param.ToString();

        SVector<aclTensor *> inTensors{x, y, plan.gatherAclOffset};
        SVector<aclTensor *> outTensors{A};
        Status status = RunAsdOpsV2(plan.GetStream(), opDesc, inTensors, outTensors, plan.GetWorkspace());
        ASDSIP_ECHECK(status.Ok(), status.Message(), ErrorType::ACL_ERROR_INTERNAL_ERROR);

        ASDSIP_LOG(INFO) << "Execute asdBlasCgerc success.";
        return ErrorType::ACL_SUCCESS;
    } catch (std::bad_cast &e) {
        // Handle the op cgerc exception
        ASDSIP_ELOG(ErrorType::ACL_ERROR_INTERNAL_ERROR) << "Cgerc Error: " << e.what();
        return ErrorType::ACL_ERROR_INTERNAL_ERROR;
    }
}

AspbStatus asdBlasMakeCgercPlan(asdBlasHandle handle)
{
    std::lock_guard<std::mutex> lock(blas_mtx);
    ASDSIP_ECHECK(handle != nullptr, "blas MakeCgercPlan Fail.", ErrorType::ACL_ERROR_INTERNAL_ERROR);

    AsdSip::BlasCgercPlan *plan = nullptr;
    try {
        plan = new AsdSip::BlasCgercPlan();
        BlasPlanCache::MakePlan(handle, plan);
    } catch (const std::exception &e) {
        if (plan != nullptr) {
            delete plan;
        }
        delete static_cast<int *>(handle);
        ASDSIP_ELOG(ErrorType::ACL_ERROR_INTERNAL_ERROR) << "Make Cgerc Plan failed: " << e.what();
        throw std::runtime_error("Make Cgerc Plan failed.");
    }
    if (plan->CreateTensor() != ErrorType::ACL_SUCCESS) {
        plan->MarkFailed();
        ASDSIP_ELOG(ErrorType::ACL_ERROR_INTERNAL_ERROR) << "Fail to create blas asdBlasMakeCgercPlan mask tensor.";
        return ErrorType::ACL_ERROR_INTERNAL_ERROR;
    }
    plan->MarkInitialized();
    return ErrorType::ACL_SUCCESS;
}
}  // namespace AsdSip

File: sip_pta/test/blas/cgerc.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Copyright (c) 2026 Huawei Technologies Co., Ltd.
This program is free software, you can redistribute it and/or modify it under the terms and conditions of
CANN Open Software License Agreement Version 2.0 (the "License").
Please refer to the License for details. You may not use this file except in compliance with the License.
THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
See LICENSE in the root of the software repository for the full text of the License.
"""

import logging
import sys
from dataclasses import dataclass
from typing import Tuple, Union

import torch
import torch_npu
import torch_sip

# 配置全局日志 (G.LOG.02)
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[logging.StreamHandler(sys.stdout)]
)
logger = logging.getLogger("torch_sip_cgerc_test")


@dataclass(frozen=True)
class GercConfig:
    """参数对象：封装 CGERC 的所有维度与标量配置"""
    m: int
    n: int
    alpha: Union[complex, float]
    name: str = "Case"


class CgercTester:
    """CGERC 算子功能测试类"""

    def __init__(self, device: str = "npu:0"):
        self.device = device

    @staticmethod
    def to_column_major(tensor: torch.Tensor) -> torch.Tensor:
        """确保矩阵满足列优先布局需求 (静态方法)"""
        if tensor.dim() == 2:
            return tensor.t().contiguous().t()
        return tensor.contiguous()

    def get_complex_tensor(self, shape: Tuple[int, ...]) -> torch.Tensor:
        """安全生成复数张量 (先 CPU 后搬移，避开 NPU 随机数限制)"""
        real = torch.randn(shape, dtype=torch.float32)
        imag = torch.randn(shape, dtype=torch.float32)
        return torch.complex(real, imag).to(self.device)

    def run_case(self, cfg: GercConfig) -> bool:
        """
        执行单次测试用例。
        G.ERR.01: 最小化 try 块。
        """
        # 1. 准备数据与输入
        mat_a = self.get_complex_tensor((cfg.m, cfg.n))
        vec_x = self.get_complex_tensor((cfg.m,))
        vec_y = self.get_complex_tensor((cfg.n,))

        # 备份初始数据用于 CPU 参考计算
        mat_a_ref = mat_a.cpu()
        vec_x_cpu = vec_x.cpu()
        vec_y_cpu = vec_y.cpu()

        a_in = mat_a.contiguous()
        x_in = vec_x.contiguous()
        y_in = vec_y.contiguous()

        # 2. 核心调用 (G.ERR.01: 仅包裹算子核心逻辑)
        try:
            torch_sip.asd_blas_cgerc(a_in, x_in, y_in, cfg.alpha)
        except Exception as exc:
            logger.error("[%s] 算子执行崩溃: %s", cfg.name, exc)
            return False

        # 3. CPU 参考计算: A = alpha * outer(x, conj(y)) + A
        ref = (cfg.alpha * torch.outer(vec_x_cpu, vec_y_cpu.conj())) + mat_a_ref

        # 4. 比对验证
        is_close = torch.allclose(a_in.cpu(), ref, rtol=1e-4, atol=1e-4)

        status = "PASS" if is_close else "FAIL"
        logger.info("[%s] %-20s | M=%d, N=%d | Alpha=%s",
                    status, cfg.name, cfg.m, cfg.n, cfg.alpha)

        if not is_close:
            max_diff = torch.abs(a_in.cpu() - ref).max()
            logger.error("      Max Diff: %.6e", max_diff)

        return is_close


def main():
    """主测试套件入口"""
    tester = CgercTester()
    logger.info("开始 CGERC 专项测试 ...\n")

    # 使用参数对象定义测试套件
    test_suites = [
        GercConfig(8, 4, 1.0, "Real_Alpha"),
        GercConfig(16, 16, complex(0.5, 0.5), "Complex_Alpha"),
        GercConfig(128, 64, 1.0, "Large_Rectangle"),
    ]

    all_passed = True
    for config in test_suites:
        if not tester.run_case(config):
            all_passed = False

    logger.info("-" * 60)
    if all_passed:
        logger.info("测试结论: ✅ 全部通过")
    else:
        logger.error("测试结论: ❌ 存在失败项")

    return 0 if all_passed else 1


if __name__ == "__main__":
    # 整个脚本唯一的 sys.exit 出口
    sys.exit(main())
